core/trino-main/src/main/java/io/trino/operator/DuplicateRowFinder.java

```diff
@@ -0,0 +1,97 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.spi.Page;
+import io.trino.spi.TrinoException;
+import io.trino.spi.block.Block;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.type.Type;
+import io.trino.type.BlockTypeOperators;
+import io.trino.type.BlockTypeOperators.BlockPositionIsDistinctFrom;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.trino.spi.StandardErrorCode.MERGE_TARGET_ROW_MULTIPLE_MATCHES;
+import static io.trino.spi.connector.MergeDetails.DEFAULT_CASE_OPERATION_NUMBER;
+import static io.trino.spi.connector.MergeDetails.INSERT_OPERATION_NUMBER;
+import static io.trino.spi.type.IntegerType.INTEGER;
+
+/**
+ * This class throws an exception if two adjacent rows have the same identity.
+ * It compares both the rowId column as well as the writeRedistributionColumns,
+ * because it is called before the page is separated by partition, and
+ * different partitions can have identical rowId columns.
+ */
+public class DuplicateRowFinder
+{
+    private final List<Integer> channels;
+    private final List<BlockPositionIsDistinctFrom> channelPositionIsDistinctFrom;
+    private Page lastPage;
+    private int lastRowIdPosition = -1;
+
+    public DuplicateRowFinder(List<ColumnHandle> dataColumns, List<Type> dataColumnTypes, List<ColumnHandle> writeRedistributionColumns, Type rowIdType)
+    {
+        BlockTypeOperators blockTypeOperators = new BlockTypeOperators();
+        Map<ColumnHandle, Type> typeMap = IntStream.range(0, dataColumns.size())
+                .boxed()
+                .collect(toImmutableMap(dataColumns::get, dataColumnTypes::get));
+        ImmutableList.Builder<Integer> channelsBuilder = ImmutableList.builder();
+        ImmutableList.Builder<BlockPositionIsDistinctFrom> channelPositionIsDistinctFromBuilder = ImmutableList.builder();
+        for (int channel = 0; channel < writeRedistributionColumns.size(); channel++) {
+            channelsBuilder.add(channel);
+            ColumnHandle handle = writeRedistributionColumns.get(channel);
+            channelPositionIsDistinctFromBuilder.add(blockTypeOperators.getDistinctFromOperator(typeMap.get(handle)));
+        }
+        channelsBuilder.add(writeRedistributionColumns.size());
+        channelPositionIsDistinctFromBuilder.add(blockTypeOperators.getDistinctFromOperator(rowIdType));
+        channels = channelsBuilder.build();
+        channelPositionIsDistinctFrom = channelPositionIsDistinctFromBuilder.build();
+    }
+
+    /**
+     * This method looks for sequential duplicates in the target rowId block and redistribution column blocks.
+     * A sequential duplicate signals that multiple target table rows matched a source row, which violates the
```

Comment:
I am not sure the assumption about sequential occurrences is enforced now.

```
trino:default> CREATE TABLE region WITH (transactional=true) AS TABLE tpch.tiny.region;
CREATE TABLE: 5 rows
```
```
trino:default> EXPLAIN MERGE INTO region r USING tpch.tiny.nation n ON r.regionkey = n.nationkey WHEN MATCHED THEN UPDATE SET comment = n.comment;
                                                                                                                                                                 Query Plan
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Fragment 0 [COORDINATOR_ONLY]
     Output layout: [rows]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     Output[rows]
     │   Layout: [rows:bigint]
     │   Estimates: {rows: ? (?), cpu: ?, memory: ?, network: ?}
     └─ TableCommit[hive:default:region]
        │   Layout: [rows:bigint]
        └─ LocalExchange[SINGLE] ()
           │   Layout: [partialrows:bigint, fragment:varbinary]
           │   Estimates: {rows: ? (?), cpu: ?, memory: ?, network: ?}
           └─ RemoteSource[1]
                  Layout: [partialrows:bigint, fragment:varbinary]

 Fragment 1 [ROUND_ROBIN]
     Output layout: [partialrows, fragment]
     Output partitioning: SINGLE []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     Merge[hive:default:region]
     │   Layout: [partialrows:bigint, fragment:varbinary]
     └─ LocalExchange[SINGLE] ()
        │   Layout: [regionkey:bigint, name:varchar(25), comment:varchar(152), operation:integer, field:row(originalTransaction bigint, bucket integer, rowId bigint)]
        │   Estimates: {rows: ? (?), cpu: ?, memory: ?, network: ?}
        └─ RemoteSource[2]
               Layout: [regionkey:bigint, name:varchar(25), comment:varchar(152), operation:integer, field:row(originalTransaction bigint, bucket integer, rowId bigint)]

 Fragment 2 [SOURCE]
     Output layout: [regionkey, name, comment, operation, field]
     Output partitioning: ROUND_ROBIN []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     DeleteAndInsert
     │   Layout: [regionkey:bigint, name:varchar(25), comment:varchar(152), operation:integer, field:row(originalTransaction bigint, bucket integer, rowId bigint)]
     └─ Project[]
        │   Layout: [field:row(originalTransaction bigint, bucket integer, rowId bigint), merge_row:row(regionkey bigint, name varchar(25), comment varchar(152), integer, integer)]
        │   Estimates: {rows: ? (?), cpu: ?, memory: 650B, network: 650B}
        │   merge_row := CAST((CASE WHEN "present" THEN ROW ("regionkey", "name", "comment_2", (NOT ("present" IS NULL)), 3) ELSE ROW (CAST(null AS bigint), CAST(null AS varchar(25)), CAST(null AS varchar(152)), (NOT ("present" IS NULL)), -1) END) AS ROW(regionkey bigint, n
        └─ LeftJoin[("nationkey" = "regionkey")][$hashvalue, $hashvalue_3]
           │   Layout: [comment_2:varchar(152), field:row(originalTransaction bigint, bucket integer, rowId bigint), regionkey:bigint, name:varchar(25), present:boolean]
           │   Estimates: {rows: ? (?), cpu: ?, memory: 650B, network: 650B}
           │   Distribution: REPLICATED
           ├─ ScanProject[table = tpch:nation:sf0.01, grouped = false]
           │      Layout: [nationkey:bigint, comment_2:varchar(152), $hashvalue:bigint]
           │      Estimates: {rows: 25 (2.38kB), cpu: 2.16k, memory: 0B, network: 0B}/{rows: 25 (2.38kB), cpu: 4.53k, memory: 0B, network: 0B}
           │      $hashvalue := combine_hash(bigint '0', COALESCE("$operator$hash_code"("nationkey"), 0))
           │      nationkey := tpch:nationkey
           │      comment_2 := tpch:comment
           └─ LocalExchange[HASH][$hashvalue_3] ("regionkey")
              │   Layout: [field:row(originalTransaction bigint, bucket integer, rowId bigint), regionkey:bigint, name:varchar(25), present:boolean, $hashvalue_3:bigint]
              │   Estimates: {rows: 5 (650B), cpu: 1.85k, memory: 0B, network: 650B}
              └─ RemoteSource[3]
                     Layout: [field:row(originalTransaction bigint, bucket integer, rowId bigint), regionkey:bigint, name:varchar(25), present:boolean, $hashvalue_4:bigint]

 Fragment 3 [SOURCE]
     Output layout: [field, regionkey, name, present, $hashvalue_5]
     Output partitioning: BROADCAST []
     Stage Execution Strategy: UNGROUPED_EXECUTION
     ScanProject[table = hive:default:region, grouped = false]
         Layout: [field:row(originalTransaction bigint, bucket integer, rowId bigint), regionkey:bigint, name:varchar(25), present:boolean, $hashvalue_5:bigint]
         Estimates: {rows: 5 (650B), cpu: 595, memory: 0B, network: 0B}/{rows: 5 (650B), cpu: 1.22k, memory: 0B, network: 0B}
         present := true
         $hashvalue_5 := combine_hash(bigint '0', COALESCE("$operator$hash_code"("regionkey"), 0))
         field := $row_id:struct<originalTransaction:bigint,bucket:int,rowId:bigint>:SYNTHESIZED
         regionkey := regionkey:bigint:REGULAR
         name := name:varchar(25):REGULAR


(1 row)
```

- The initial right join got rotated to LeftJoin, so the target rows are no longer guaranteed to be next to each other.
- the `Merge` plan node is preceded with `Output partitioning: ROUND_ROBIN []`, which scatters the rows further.

Reply:
You were correct, the constraint wasn't previously enforced because the mechanism to find the appropriate `TableScanNode` would find the wrong one if the `JOIN` got flipped.

The constraint is now enforced, and as detailed in https://github.com/trinodb/trino/pull/7933#issuecomment-864418393, using the `updateTarget` mechanism that guarantees that the right `TableScanNode` will be found no matter where it is in the tree.